### PR TITLE
boost minibatches

### PR DESCRIPTION
**Pull request by ferrine**
- [x] support multidimentional subsampling #2125

- [x] add minibatch inteface #2056

- [x] tests

- [x] notebooks

### CODE REVIEW

File: pymc3/model.py [hunk 1]
```diff
@@ -914,7 +938,7 @@ def __init__(self, type=None, owner=None, index=None, name=None, data=None,
             data = pandas_to_array(data)
             type = TensorType(distribution.dtype, data.shape)
 
-        super(TensorVariable, self).__init__(type, None, None, name)
+        super(ObservedRV, self).__init__(type, owner, index, name)
```
Comment (twiecki) [inline pymc3/model.py]: Why is it calling that super class?

Comment (ferrine) [inline pymc3/model.py]: Consistent implementation

Comment (twiecki) [inline pymc3/model.py]: But it doesn't even inherit from ObservedRV or am I missing that?

Comment (ferrine) [inline pymc3/model.py]: Oops

Comment (ferrine) [inline pymc3/model.py]: Oops

Comment (ferrine) [inline pymc3/model.py]: Stop, that's ok. `super` is right for that we call it on our class, not on base classes

Comment (twiecki) [inline pymc3/model.py]: It should be `FreeRV`.

Comment (ferrine) [inline pymc3/model.py]: See http://stackoverflow.com/a/576183

Comment (ferrine) [inline pymc3/model.py]: My fault:))

Comment (ferrine) [inline pymc3/model.py]: Seems like all is OK https://github.com/ferrine/pymc3/blob/90ab385b48fa4ab9e58901befd86b25af00f1218/pymc3/model.py#L941

Comment (twiecki) [inline pymc3/model.py]: It should definitely be `FreeRV`.

**Comment by twiecki** [comment]
@ferrine One check-box above is unchecked, do you consider this done?

File: pymc3/model.py (hunk 1 above)
Comment (twiecki) [inline pymc3/model.py]: You're correct.

**Comment by twiecki** [comment]
@ferrine is this good from your end?

**Comment by ferrine** [comment]
I need tests for minibatch class

**Comment by fonnesbeck** [comment]
Would be great to convert the notebooks that use the old minibatch interface over to this one as part of the PR.

**Comment by ferrine** [comment]
Once tests pass I'll cherry pick #2097

**Comment by fonnesbeck** [comment]
So, does this replace `DataSampler`?

**Comment by ferrine** [comment]
Yes. This implementation is much more faster in runtime.

**Comment by ferrine** [comment]
`DataSampler` is used only in tests. I think we are free to remove it before #2223

**Comment by ferrine** [comment]
Previously I was suggested to change OOP inference to pm.fit. There is OOP inference in quick start api. I see it suitable for VI notebooks. So I am a bit confused why should we hide it in examples. It is very handy.
Thus I see further work in introducing new minibatches. Any other thoughts?

**Comment by fonnesbeck** [comment]
@ferrine it's a matter of having consistent primary interfaces across PyMC3. I'm fine with having convenient methods for advanced users, but I think we need to be showing the same approach to fitting models among the suite of notebooks that we maintain. If we think OO is the way to go, we can have that discussion and change it across the board, including for the MCMC classes. 

That said, I'm not opposed to having a notebook of advanced features that could show off a comprehensive example that uses all the handy shortcuts that you have made available.

**Comment by ferrine** [comment]
Hm, seems like very comprehensive `__init__` and `clone` method can be together

File: pymc3/data.py [hunk 2]
```diff
@@ -90,59 +90,258 @@ def __hash__(self):
...
+    Notable thing is that :class:`Minibatch` has `shared`, `minibatch`, attributes
+    you can call later
+    >>> x.set_value(np.random.laplace(size=(100, 100)))
+
+    and minibatches will be then from new storage
+    it directly affects `x.shared`.
+    the same thing would be but less convenient
+    >>> x.shared.set_value(pm.floatX(np.random.laplace(size=(100, 100))))
+    
+    programmatic way to change storage is as following
```
Comment (fonnesbeck) [inline pymc3/data.py]: as follows

File: pymc3/data.py [hunk 3]
```diff
@@ -90,59 +90,258 @@ def __hash__(self):
...
+
+    Examples
+    --------
+    Consider we have data
+    >>> data = np.random.rand(100, 100)
+
+    if we want 1d slice of size 10 we do
+    >>> x = Minibatch(data, batch_size=10)
+    
+    Note, that your data is casted to `floatX` if it is not integer type
```
Comment (fonnesbeck) [inline pymc3/data.py]: is cast

File: pymc3/data.py [hunk 4]
```diff
@@ -90,59 +90,258 @@ def __hash__(self):
...
-    <pymc3.variational.approximations.MeanField object at 0x...>
-    >>> new.sample_vp(draws=1000)['mu'].mean()
-    10.08339999101371
-    >>> new.sample_vp(draws=1000)['sd'].mean()
-    1.2178044136104513
+    >>> datagen = partial(np.random.laplace, size=(100, 100))
+    >>> x = Minibatch(datagen(), batch_size=100, update_shared_f=datagen)
+    >>> x.update_shared()
+
+    To be more precise of how we get minibatch, here is a demo
```
Comment (fonnesbeck) [inline pymc3/data.py]: To be more concrete about how we get ...

File: pymc3/data.py [hunk 5]
```diff
@@ -90,59 +90,258 @@ def __hash__(self):
...
+    2) create random slice of size 10
+    >>> ridx = pm.tt_rng().uniform(size=(10,), low=0, high=data.shape[0]-1e-10).astype('int64')
+
+    3) take that slice
+    >>> minibatch = shared[ridx]
+
+    That's done. So if you'll need some replacements in the graph 
+    >>> testdata = pm.floatX(np.random.laplace(size=(1000, 10)))
+
+    you are free to use a kind of this one as `x` is regular Theano Tensor
```
Comment (fonnesbeck) [inline pymc3/data.py]: (I don't quite understand this sentence)

File: pymc3/data.py [hunk 6]
```diff
@@ -90,59 +90,258 @@ def __hash__(self):
...
+    >>> node = x ** 2  # arbitrary expressions
+    >>> rnode = theano.clone(node, replacements)
+    >>> assert (testdata ** 2 == rnode.eval()).all()
+
+    For more complex slices some more code is needed that can seem not so clear
+    They are
+    >>> moredata = np.random.rand(10, 20, 30, 40, 50)
+    
+    default total_size is then (10, 20, 30, 40, 50) but 
+    can be less verbose in sove cases
```
Comment (fonnesbeck) [inline pymc3/data.py]: some cases

**Comment by fonnesbeck** [comment]
The down side to having the batches as tensors is that we cannot iterate over them, which is sometimes necessary. You can't even pass them to `scan` as a sequence.

**Comment by ferrine** [comment]
@fonnesbeck thanks for reviewing this:)
Case of iterating minibatches is not covered ofc. But `scan` is ok. You can do arbitrary theano operations over this minibatch. It is sampled once per graph execution

**Comment by ferrine** [comment]
I think that's done after rebase, feel free to leave feedback about refactored notebooks

**Comment by ferrine** [comment]
Need review

**Comment by taku-y** [comment]
In convolutional_vae_keras_advi.ipynb `from pymc3.variational import advi_minibatch` is not used.

**Comment by taku-y** [comment]
LDA notebook explains that "is automativally exponentiated (thus bounded to be positive) in advi_minibatch(), the estimation function.". advi_minibatch() should be removed, or readers might be confused.

**Comment by junpenglao** [comment]
Did you also update `test_advi.py`? It's still testing the old advi api.

**Comment by junpenglao** [comment]
Oh I see there is `test_variational_inference.py` for the new api

File: pymc3/tests/conftest.py [hunk 7]
```diff
@@ -1,7 +1,32 @@
...
+        self.rng = np.random.RandomState(random_seed)
+        self.data = data
+        self.n = batchsize
+
+    def __iter__(self):
+        return self
+
+    def __next__(self):
+        idx = (self.rng
+               .uniform(size=self.n,
```
Comment (taku-y) [inline pymc3/tests/conftest.py:21]: You didn't use `randint()` because Theano doesn't support `randint()`, right?

**Comment by taku-y** [comment]
I had a close look at this PR and I believe it's ready for merge. Although some tests failed but those doesn't seem to be related to the PR. Actually, `test_minibatches.py` passed with 100%. I want to merge it unless I hear otherwise.

File: pymc3/tests/conftest.py (hunk 7 above)
Comment (ferrine) [reply pymc3/tests/conftest.py:21]: Yes, that is it

Comment (ferrine) [reply pymc3/tests/conftest.py:21]: But here I've left it just for test purposes. I use numpy for this test

Comment (taku-y) [reply pymc3/tests/conftest.py:21]: It might be a reasonable option wrt consistency between numpy and theano.

**Comment by pwl** [comment]
I'm getting dimension mismatch with the following model:
```python
data = np.random.rand(101, 202)
data_t = pm.Minibatch(data, [10,11])
with pm.Model() as model:
    U = pm.Normal('U', 0, 1, shape=(101, 3))
    V = pm.Normal('V', 0, 1, shape=(202, 3))
    reads = pm.Normal('reads', mu=pm.math.dot(U,V.T), observed=data_t,
                      total_size=data.shape)
```

```
ValueError: Input dimension mis-match. (input[0].shape[0] = 10, input[1].shape[0] = 101)
```

but it works when I replace `data_t` with `data`.

**Comment by junpenglao** [comment]
@pwl I think the correct usage goes something like this?
```python
data = np.random.rand(101, 202)
data_t = pm.Minibatch(data, [10,11])
with pm.Model() as model:
    U = pm.Normal('U', 0, 1, shape=(10, 3), total_size=(101, 3))
    V = pm.Normal('V', 0, 1, shape=(11, 3), total_size=(202, 3))
    reads = pm.Normal('reads', mu=pm.math.dot(U,V.T), observed=data_t,
                      total_size=data.shape)
```

**Comment by pwl** [comment]
@junpenglao thanks! I was just testing that, it seems that it works.  How can I extract the full extent of `U` and `V` after fitting the model?  `sample` is returning just the minibatch-sized chunk of `U`.

EDIT: just to clarify, I'm doing
```python
data = np.random.rand(101, 202)
data_t = pm.Minibatch(data, [10,11])
with pm.Model() as model:
    U = pm.Normal('U', 0, 1, shape=(10, 3), total_size=(101, 3))
    V = pm.Normal('V', 0, 1, shape=(11, 3), total_size=(202, 3))
    reads = pm.Normal('reads', mu=pm.math.dot(U,V.T), observed=data_t,
                      total_size=data.shape)
    advi = pm.ADVI()
    approx = advi.fit(1)
    trace = approx.sample(1)
trace['U'].shape # gives (1, 10, 3)
```

**Comment by junpenglao** [comment]
@pwl I am not sure, this is a brand new feature - I havent got a chance to play with it yet. @ferrine?

**Comment by pwl** [comment]
Ok, I think I misunderstood the interface and somehow ignored the need to add indices by hand as in
```python
n_i = 101
n_j = 202
data = np.random.rand(n_i, n_j)
data_t = pm.Minibatch(data, [10,11])
data_i_idx = pm.Minibatch(range(n_i),10)
data_j_idx = pm.Minibatch(range(n_j),11)
with pm.Model() as model:
    U = pm.Normal('U', 0, 1, shape=(n_i, 3))[data_i_idx,:]
    V = pm.Normal('V', 0, 1, shape=(n_j, 3))[data_j_idx,:]
    reads = pm.Normal('reads', mu=pm.math.dot(U,V.T), observed=data_t,
                      total_size=data.shape)
    advi = pm.ADVI()
    approx = advi.fit(10000)
    trace = approx.sample(100)
trace['U'].shape # works fine now: (100, 101, 3)
```
I had an impression that the indexing will be done automatically under the hood.

**Comment by pwl** [comment]
@ferrine thanks for this great PR, this really simplifies and unifies the minibatch training!

**Comment by ferrine** [comment]
@pwl I would also recommend setting different random seeds on different dimensions. Orelse you get correlated observations. You can find reference in the docstring

**Comment by ferrine** [comment]
@junpenglao I haven't yet played with multidim training too. But for 1d I found this very convenient

**Comment by junpenglao** [comment]
@ferrine We should extend the probabilistic matrix factorization example to demonstrate the multi-dimensional mini batch training.

**Comment by ferrine** [comment]
Yes that is a good idea

**Comment by mbabadi** [comment]
@ferrine thanks for the minibatch implementation -- great PR. I am not quite familiar with the guts of PyMC3 yet, but looking at @pwl's matrix factorization model, I wonder whether there is a _guarantee_ that `data_t`, `data_i_idx` and `data_j_idx` are evaluated only once during each round of gradient/ELBO evaluation? if one of them falls behind (or moves forward), the game is ruined.

**Comment by ferrine** [comment]
Hi, 
Random indices are evaluated once per function evaluation by design of theano. What about manual minibatch.eval() or independent usage within several functions that's the thing I don't know how to control. That is what user should take care of. Personally I prefer putting minibatches in inference as a replacement, less error prone.

**Comment by mbabadi** [comment]
Thank you very much for your comment @ferrine.

I wonder, perhaps it would be _even more_ error proof if it was possible to extract the slicing indices from the data minibatch directly. @pwl's solution involving multiple minibatches on linear ranges for each dimension makes room for strange bugs down the road (what if a future version of theano decides to evaluate some tensors twice?)

Is there an elegant solution to having a tuple of integer 1D tensors inside the minibatch class, one for each dimension, that would always be in sync with the slice `minibatch.eval()` would yield, and that the user could use/evaluate without the side effect of changing the state of the RNG?

**Comment by ferrine** [comment]
Having that on instance level is not a solution since you have different minibatches for different tensors. Class level solution is needed here. I see it is not easy to implement, and as you have a workaround it does not worth at all.

**Pull request merged**